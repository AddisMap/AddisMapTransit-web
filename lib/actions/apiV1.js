import { createAction } from 'redux-actions'
import L from 'leaflet'

import {
  bikeRentalQuery,
  createQueryAction,
  findGeometryForTrip,
  findNearbyStops,
  findRoutesAtStop,
  findStopError,
  findStopResponse,
  findStopsForTrip,
  findStopTimesForStop,
  findStopTimesForTrip,
  findTripError,
  findTripResponse,
  parkAndRideQuery,
  vehicleRentalQuery
} from './api'
import { zoomToStop } from './map'

const findTrip = (params) =>
  createQueryAction(
    `index/trips/${params.tripId}`,
    findTripResponse,
    findTripError,
    {
      noThrottle: true,
      postprocess: (payload, dispatch) => {
        dispatch(findStopsForTrip({ tripId: params.tripId }))
        dispatch(findStopTimesForTrip({ tripId: params.tripId }))
        dispatch(findGeometryForTrip({ tripId: params.tripId }))
      }
    }
  )

export const findNearbyAmenitiesResponse = createAction(
  'FIND_NEARBY_AMENITIES_RESPONSE'
)
export const findNearbyAmenitiesError = createAction(
  'FIND_NEARBY_AMENITIES_ERROR'
)

function findNearbyAmenities(params, stopId) {
  return function (dispatch, getState) {
    const { lat, lon, radius } = params
    const bounds = L.latLng(lat, lon).toBounds(radius)
    const { lat: low, lng: left } = bounds.getSouthWest()
    const { lat: up, lng: right } = bounds.getNorthEast()
    dispatch(
      parkAndRideQuery(
        { lowerLeft: `${low},${left}`, upperRight: `${up},${right}` },
        findNearbyAmenitiesResponse,
        findNearbyAmenitiesError,
        {
          noThrottle: true,
          rewritePayload: (payload) => {
            return {
              parkAndRideLocations: payload,
              stopId
            }
          }
        }
      )
    )
    dispatch(
      bikeRentalQuery(
        { lowerLeft: `${low},${left}`, upperRight: `${up},${right}` },
        findNearbyAmenitiesResponse,
        findNearbyAmenitiesError,
        {
          rewritePayload: (payload) => {
            return {
              bikeRental: payload,
              stopId
            }
          }
        }
      )
    )
    dispatch(
      vehicleRentalQuery(
        { lowerLeft: `${low},${left}`, upperRight: `${up},${right}` },
        findNearbyAmenitiesResponse,
        findNearbyAmenitiesError,
        {
          rewritePayload: (payload) => {
            return {
              stopId,
              vehicleRental: payload
            }
          }
        }
      )
    )
  }
}

export function findStop(params) {
  return createQueryAction(
    `index/stops/${params.stopId}`,
    findStopResponse,
    findStopError,
    {
      noThrottle: true,
      postprocess: (payload, dispatch) => {
        dispatch(findRoutesAtStop(params.stopId))
        dispatch(findStopTimesForStop(params))
      },
      serviceId: 'stops'
    }
  )
}

const fetchStopInfo = (stop) =>
  async function (dispatch, getState) {
    await dispatch(findStop({ stopId: stop.stopId }))
    const state = getState()
    const { nearbyRadius: radius } = state.otp.config.stopViewer
    const fetchedStop = state.otp.transitIndex.stops?.[stop?.stopId]
    // TODO: stop not found message
    if (!fetchedStop) return

    const { lat, lon } = fetchedStop
    if (radius > 0) {
      dispatch(
        findNearbyStops(
          {
            includeRoutes: true,
            includeStopTimes: true,
            lat,
            lon,
            radius
          },
          stop.stopId
        )
      )
      dispatch(findNearbyAmenities({ lat, lon, radius }, stop.stopId))
      dispatch(zoomToStop(fetchedStop))
    }
  }

export default { fetchStopInfo, findTrip }
