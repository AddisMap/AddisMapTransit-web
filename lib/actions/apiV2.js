import {
  createQueryAction,
  findGeometryForTrip,
  findStopError,
  findStopResponse,
  findStopTimesForStop,
  findStopTimesForTrip,
  findTripError,
  findTripResponse
} from './api'
import { setMapZoom } from './config'
import { zoomToStop } from './map'

/**
 * Generic helper for crafting GraphQL queries.
 */
function createGraphQLQueryAction(
  query,
  variables,
  responseAction,
  errorAction,
  options
) {
  const endpoint = 'index/graphql'
  const fetchOptions = {
    body: JSON.stringify({ query, variables }),
    headers: { 'Content-Type': 'application/json' },
    method: 'POST'
  }
  return createQueryAction(endpoint, responseAction, errorAction, {
    ...options,
    fetchOptions
  })
}

const findTrip = (params) =>
  createGraphQLQueryAction(
    `{
        trip(id: "${params.tripId}") {
          id: gtfsId
          route {
            id: gtfsId
            agency {
              id: gtfsId
              name
              url
              timezone
              lang
              phone
              fareUrl
            }
            shortName
            longName
            type
            url
            color
            textColor
            routeBikesAllowed: bikesAllowed
            bikesAllowed
          }
          serviceId
          tripHeadsign
          directionId
          blockId
          shapeId
          wheelchairAccessible
          bikesAllowed
          tripBikesAllowed: bikesAllowed

          stops {
			gtfsId: id
			code
			name
			lat
			lon			
		}
        }
      }`,
    {},
    findTripResponse,
    findTripError,
    {
      noThrottle: true,
      postprocess: (payload, dispatch) => {
        // FIXME: integrate into graphql request
        dispatch(findStopTimesForTrip({ tripId: params.tripId }))
        dispatch(findGeometryForTrip({ tripId: params.tripId }))
      },
      rewritePayload: (payload) => {
        return payload?.data?.trip || {}
      }
    }
  )

const fetchStopInfo = (stop) => {
  const { stopId } = stop
  if (!stopId) return {}

  return createGraphQLQueryAction(
    `{
    stop(id: "${stopId}") {
      id: gtfsId
      lat
      lon
      locationType
      name
      wheelchairBoarding
      zoneId
      geometries {
        geoJson
      }
      routes {
        id: gtfsId
        agency {
          gtfsId
          name
        }
        longName
        mode
        color
        shortName
      }
    }
  }
`,
    {},
    findStopResponse,
    findStopError,
    {
      noThrottle: true,
      postprocess: (payload, dispatch) => {
        if (payload.errors) {
          return findStopError(payload.errors)
        }
        const { stop } = payload?.data
        if (!stop) return findStopError()

        // TODO: nearby stops and amenities
        dispatch(findStopTimesForStop({ stopId: stop.id }))
        dispatch(zoomToStop(stop))

        if (stop?.geometries?.geoJson?.type !== 'Point') {
          dispatch(setMapZoom(10))
        }
      },
      rewritePayload: (payload) => {
        const { stop } = payload?.data
        if (!stop) return findStopError()

        const color = stop.routes?.length > 0 && `#${stop.routes[0].color}`

        // Doing some OTP1 compatibility rewriting here
        return {
          ...stop,
          agencyId: stop?.agency?.gtfsId,
          agencyName: stop?.agency?.name,
          color
        }
      },
      serviceId: 'stops'
    }
  )
}

export default { fetchStopInfo, findTrip }
