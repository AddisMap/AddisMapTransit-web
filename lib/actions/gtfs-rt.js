import { createAction } from 'redux-actions'
if (typeof (fetch) === 'undefined') require('isomorphic-fetch')

export const receivedGtfsRtVehiclePositionsError = createAction('GTFS_RT_VEHICLE_POSITIONS_ERROR')
export const receivedGtfsRtVehiclePositionsResponse = createAction('GTFS_RT_VEHICLE_POSITIONS_RESPONSE')
export const requestGtfsRtVehiclePositionsResponse = createAction('GTFS_RT_VEHICLE_POSITIONS_REQUEST')

export function gtfsRtVehiclePositionsQuery (url) {
  return async function (dispatch, getState) {
    dispatch(requestGtfsRtVehiclePositionsResponse())
    let json
    try {
      const response = await fetch(url)
      if (response.status >= 400) {
        const error = new Error('Received error from server')
        error.response = response
        throw error
      }
      json = await response.json()
    } catch (err) {
      return dispatch(receivedGtfsRtVehiclePositionsError(err))
    }

    dispatch(receivedGtfsRtVehiclePositionsResponse({ data: json, feedId: url }))
  }
}
