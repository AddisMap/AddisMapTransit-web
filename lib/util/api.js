if (typeof (fetch) === 'undefined') require('isomorphic-fetch')

/**
 * Gets JSON from a response object.
 * If the response contains an error code, logs the error in the console and throw exception.
 * @param {*} res The HTTP response object.
 */
export function getJsonAndCheckResponse (res) {
  if (res.status >= 400) {
    const error = new Error('Received error from server')
    error.response = res
    throw error
  }
  return res.json()
}
