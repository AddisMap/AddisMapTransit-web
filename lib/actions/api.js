/* globals fetch */

import { createAction } from 'redux-actions'
import qs from 'qs'

if (typeof (fetch) === 'undefined') {
  require('isomorphic-fetch')
}

import { queryIsValid } from '../util/state'

import queryParams from '../util/query-params'
import { filterProfileOptions } from '../util/profile'

// Generic API actions

export const routingRequest = createAction('ROUTING_REQUEST')
export const routingResponse = createAction('ROUTING_RESPONSE')
export const routingError = createAction('ROUTING_ERROR')

let lastSearchId = 0

export function routingQuery () {
  return async function (dispatch, getState) {
    const otpState = getState().otp
    const routingType = otpState.currentQuery.routingType
    const searchId = ++lastSearchId

    if (!queryIsValid(otpState)) return
    dispatch(routingRequest({ routingType, searchId }))
    const url = constructRoutingQuery(otpState)
    // setURLSearch(url)
    let json
    try {
      const response = await fetch(url)
      if (response.status >= 400) {
        const error = new Error('Received error from server')
        error.response = response
        throw error
      }
      json = await response.json()

      if (json && json.otp && json.otp.profile) {
        json.otp = filterProfileOptions(json.otp)
      }
    } catch (error) {
      return dispatch(routingError({ error, searchId }))
    }

    dispatch(routingResponse({ response: json, searchId }))
  }
}

function constructRoutingQuery (otpState) {
  const { config, currentQuery } = otpState
  const routingType = currentQuery.routingType
  const isItinerary = routingType === 'ITINERARY'
  const api = isItinerary ? config.api : config.profileApi
  const planEndpoint = `${api.host}${api.port
    ? ':' + api.port
    : ''}${api.path}/plan`

  let params = {}
  queryParams
    .filter(
      qp => qp.routingTypes.indexOf(routingType) !== -1 && qp.name in currentQuery
    )
    .forEach(qp => {
      const rewriteFunction = isItinerary
        ? qp.itineraryRewrite
        : qp.profileRewrite
      params = Object.assign(
        params,
        rewriteFunction
          ? rewriteFunction(currentQuery[qp.name])
          : { [qp.name]: currentQuery[qp.name] }
      )
    })

  const stringParams = qs.stringify(params)
  // TODO: set url hash based on params
  // setURLSearch(stringParams)
  // TODO: check that valid from/to locations are provided

  return `${planEndpoint}?${stringParams}`
}

// bike rental station query

export const receivedBikeRentalError = createAction('BIKE_RENTAL_ERROR')
export const receivedBikeRentalResponse = createAction('BIKE_RENTAL_RESPONSE')
export const requestBikeRentalResponse = createAction('BIKE_RENTAL_REQUEST')

export function bikeRentalQuery () {
  return async function (dispatch, getState) {
    const otpState = getState().otp

    dispatch(requestBikeRentalResponse())
    const api = otpState.config.api
    const planEndpoint = `${api.host}:${api.port ? ':' + api.port : ''}${api.path}/bike_rental`
    const url = planEndpoint // TODO: add bbox

    let json
    try {
      const response = await fetch(url)
      if (response.status >= 400) {
        const error = new Error('Received error from server')
        error.response = response
        throw error
      }
      json = await response.json()
    } catch (err) {
      return dispatch(receivedBikeRentalError(err))
    }

    dispatch(receivedBikeRentalResponse(json))
  }
}

export const receivedNearbyStopsResponse = createAction('NEARBY_STOPS_RESPONSE')
export const receivedNearbyStopsError = createAction('NEARBY_STOPS_ERROR')

export function findNearbyStops (params) {
  return async function (dispatch, getState) {
    const otpState = getState().otp
    const api = otpState.config.api
    params = Object.assign({ radius: 1000 }, params)
    const paramStr = Object.keys(params).map(k => `${k}=${encodeURIComponent(params[k])}`).join('&')
    const url = `${api.host}:${api.port ? ':' + api.port : ''}${api.path}/index/stops?${paramStr}`
    let stops
    try {
      const response = await fetch(url)
      if (response.status >= 400) {
        const error = new Error('Received error from server')
        error.response = response
        throw error
      }
      stops = await response.json()
    } catch (err) {
      return dispatch(receivedNearbyStopsError(err))
    }

    stops.sort((a, b) => a.dist - b.dist)
    if (stops.length > 5) stops = stops.slice(0, 5)
    dispatch(receivedNearbyStopsResponse({ stops }))

    // retrieve routes for each stop
    stops.forEach(stop => {
      dispatch(findRoutesAtStop(stop.id))
    })
  }
}

export const receivedRoutesAtStopResponse = createAction('ROUTES_AT_STOP_RESPONSE')
export const receivedRoutesAtStopError = createAction('ROUTES_AT_STOP_ERROR')

export function findRoutesAtStop (stopId) {
  return async function (dispatch, getState) {
    const otpState = getState().otp
    const api = otpState.config.api
    const url = `${api.host}:${api.port ? ':' + api.port : ''}${api.path}/index/stops/${stopId}/routes`
    let routes
    try {
      const response = await fetch(url)
      if (response.status >= 400) {
        const error = new Error('Received error from server')
        error.response = response
        throw error
      }
      routes = await response.json()
    } catch (err) {
      return dispatch(receivedRoutesAtStopError(err))
    }

    dispatch(receivedRoutesAtStopResponse({ stopId, routes }))
  }
}
