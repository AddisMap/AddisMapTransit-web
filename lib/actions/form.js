import debounce from 'lodash.debounce'
import isEqual from 'lodash.isequal'
import moment from 'moment'
import coreUtils from '@opentripplanner/core-utils'
import { createAction } from 'redux-actions'

import { queryIsValid } from '../util/state'
import {
  MobileScreens,
  setMainPanelContent,
  setMobileScreen
} from '../actions/ui'

import { routingQuery } from './api'

const {
  getDefaultQuery,
  getTripOptionsFromQuery,
  getUrlParams,
  planParamsToQuery
} = coreUtils.query

export const settingQueryParam = createAction('SET_QUERY_PARAM')
export const clearActiveSearch = createAction('CLEAR_ACTIVE_SEARCH')
export const setActiveSearch = createAction('SET_ACTIVE_SEARCH')
export const clearDefaultSettings = createAction('CLEAR_DEFAULT_SETTINGS')
export const storeDefaultSettings = createAction('STORE_DEFAULT_SETTINGS')

export function resetForm () {
  return function (dispatch, getState) {
    const otpState = getState().otp
    const { transitModes } = otpState.config.modes
    if (otpState.user.defaults) {
      dispatch(settingQueryParam(otpState.user.defaults))
    } else {
      // Get user overrides and apply to default query
      const userOverrides = coreUtils.storage.getItem('defaultQuery', {})
      const defaultQuery = Object.assign(
        getDefaultQuery(otpState.config),
        userOverrides
      )
      // Filter out non-options (i.e., date, places).
      const options = getTripOptionsFromQuery(defaultQuery)
      // Default mode is currently WALK,TRANSIT. We need to update this value
      // here to match the list of modes, otherwise the form will break.
      options.mode = ['WALK', ...transitModes.map(m => m.mode)].join(',')
      dispatch(settingQueryParam(options))
    }
  }
}

/**
 * Action to update any specified query parameter. Replaces previous series of
 * parameter-specific actions. If a search ID is provided, a routing query (OTP
 * search) will be kicked off immediately.
 */
export function setQueryParam (payload, searchId) {
  return function (dispatch, getState) {
    dispatch(settingQueryParam(payload))
    if (searchId) dispatch(routingQuery(searchId))
  }
}

// FIXME: remove
function stringToCoords(str) {
  return (str && str.split(",").map(c => +c)) || [];
}

// FIXME: remove
export function coordsToString(coords) {
  return coords.length && coords.map(c => (+c).toFixed(5)).join(", ");
}

// FIXME: remove
function parseLocationString(value) {
  const parts = value.split("::");
  const coordinates = parts[1]
    ? stringToCoords(parts[1])
    : stringToCoords(parts[0]);
  const name = parts[1] ? parts[0] : coordsToString(coordinates);
  return coordinates.length === 2
    ? {
        name: name || null,
        lat: coordinates[0] || null,
        lon: coordinates[1] || null
      }
    : null;
}

export function parseUrlQueryString (params = getUrlParams()) {
  return function (dispatch, getState) {
    // Filter out the OTP (i.e. non-UI) params and set the initial query
    const planParams = {}
    Object.keys(params).forEach(key => {
      if (!key.startsWith('ui_')) planParams[key] = params[key]
    })
    const searchId = params.ui_activeSearch || coreUtils.storage.randId()
    // Convert strings to numbers/objects and dispatch
    const parsedParams = planParamsToQuery(planParams, getState().otp.config)
    parsedParams.intermediatePlaces = planParams.intermediatePlaces
      ? planParams.intermediatePlaces.split(',').map(parseLocationString)
      : []
    dispatch(setQueryParam(parsedParams, searchId))
  }
}

let debouncedPlanTrip // store as variable here, so it can be reused.
let lastDebouncePlanTimeMs

export function formChanged (oldQuery, newQuery) {
  return function (dispatch, getState) {
    const otpState = getState().otp
    const isMobile = coreUtils.ui.isMobile()

    // If departArrive is set to 'NOW', update the query time to current
    if (otpState.currentQuery && otpState.currentQuery.departArrive === 'NOW') {
      dispatch(settingQueryParam({ time: moment().format(coreUtils.time.OTP_API_TIME_FORMAT) }))
    }

    // Determine if either from/to location has changed
    const fromChanged = !isEqual(oldQuery.from, newQuery.from)
    const toChanged = !isEqual(oldQuery.to, newQuery.to)

    // Only clear the main panel if a single location changed. This prevents
    // clearing the panel on load if the app is focused on a stop viewer but a
    // search query should also be visible.
    const oneLocationChanged = (fromChanged && !toChanged) || (!fromChanged && toChanged)
    if (oneLocationChanged) {
      dispatch(setMainPanelContent(null))
    }

    // Clear the current search and return to search screen on mobile when
    // either location changes only if not currently on welcome screen (otherwise
    // when the current position is auto-set the screen will change unexpectedly).
    if (
      isMobile &&
      (fromChanged || toChanged) &&
      otpState.ui.mobileScreen !== MobileScreens.WELCOME_SCREEN
    ) {
      dispatch(clearActiveSearch())
      dispatch(setMobileScreen(MobileScreens.SEARCH_FORM))
    }

    // Check whether a trip should be auto-replanned
    const { autoPlan, debouncePlanTimeMs } = otpState.config
    const updatePlan =
      autoPlan ||
      (!isMobile && oneLocationChanged) || // TODO: make autoplan configurable at the parameter level?
      (isMobile && fromChanged && toChanged)
    if (updatePlan && queryIsValid(otpState)) { // trip plan should be made
      // check if debouncing function needs to be (re)created
      if (!debouncedPlanTrip || lastDebouncePlanTimeMs !== debouncePlanTimeMs) {
        debouncedPlanTrip = debounce(() => dispatch(routingQuery()), debouncePlanTimeMs)
        lastDebouncePlanTimeMs = debouncePlanTimeMs
      }
      debouncedPlanTrip()
    }
  }
}
