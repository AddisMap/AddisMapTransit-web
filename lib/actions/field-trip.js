import { planParamsToQueryAsync } from '@opentripplanner/core-utils/lib/query'
import { OTP_API_DATE_FORMAT, OTP_API_TIME_FORMAT } from '@opentripplanner/core-utils/lib/time'
import moment from 'moment'
import { serialize } from 'object-to-formdata'
import qs from 'qs'
import { createAction } from 'redux-actions'

import {routingQuery} from './api'
import {toggleCallHistory} from './call-taker'
import {resetForm, setQueryParam} from './form'
import {getGroupSize, getTripFromRequest, sessionIsInvalid} from '../util/call-taker'

if (typeof (fetch) === 'undefined') require('isomorphic-fetch')

/// PRIVATE ACTIONS

const receivedFieldTrips = createAction('RECEIVED_FIELD_TRIPS')
const requestingFieldTrips = createAction('REQUESTING_FIELD_TRIPS')
const receivedFieldTripDetails = createAction('RECEIVED_FIELD_TRIP_DETAILS')
const requestingFieldTripDetails = createAction('REQUESTING_FIELD_TRIP_DETAILS')

// PUBLIC ACTIONS

export const setFieldTripFilter = createAction('SET_FIELD_TRIP_FILTER')
export const setActiveFieldTrip = createAction('SET_ACTIVE_FIELD_TRIP')
export const setGroupSize = createAction('SET_GROUP_SIZE')
export const toggleFieldTrips = createAction('TOGGLE_FIELD_TRIPS')

/**
 * Fully reset form and toggle field trips (and close call history if open).
 */
export function resetAndToggleFieldTrips () {
  return async function (dispatch, getState) {
    dispatch(resetForm(true))
    dispatch(toggleFieldTrips())
    if (getState().callTaker.callHistory.visible) dispatch(toggleCallHistory())
  }
}

/**
 * Fetch all field trip requests (as summaries).
 */
export function fetchFieldTrips () {
  return async function (dispatch, getState) {
    dispatch(requestingFieldTrips())
    const {callTaker, otp} = getState()
    if (sessionIsInvalid(callTaker.session)) return
    const {datastoreUrl} = otp.config
    const {sessionId} = callTaker.session
    let fieldTrips = []
    try {
      const res = await fetch(`${datastoreUrl}/fieldtrip/getRequestsSummary?${qs.stringify({sessionId})}`)
      fieldTrips = await res.json()
    } catch (e) {
      alert(`Error fetching field trips: ${JSON.stringify(e)}`)
    }
    dispatch(receivedFieldTrips({fieldTrips}))
  }
}

/**
 * Fetch details for a particular field trip request.
 */
export function fetchFieldTripDetails (requestId) {
  return function (dispatch, getState) {
    dispatch(requestingFieldTripDetails())
    const {callTaker, otp} = getState()
    if (sessionIsInvalid(callTaker.session)) return
    const {datastoreUrl} = otp.config
    const {sessionId} = callTaker.session
    fetch(`${datastoreUrl}/fieldtrip/getRequest?${qs.stringify({requestId, sessionId})}`)
      .then(res => res.json())
      .then(fieldTrip => dispatch(receivedFieldTripDetails({fieldTrip})))
      .catch(err => {
        alert(`Error fetching field trips: ${JSON.stringify(err)}`)
      })
  }
}

/**
 * Add note for field trip request.
 */
export function addFieldTripNote (request, note) {
  return function (dispatch, getState) {
    const {callTaker, otp} = getState()
    const {datastoreUrl} = otp.config
    if (sessionIsInvalid(callTaker.session)) return
    const {sessionId, username} = callTaker.session
    const noteData = serialize({
      sessionId,
      note: {...note, userName: username},
      requestId: request.id
    })
    return fetch(`${datastoreUrl}/fieldtrip/addNote`,
      {method: 'POST', body: noteData}
    )
      .then(() => dispatch(fetchFieldTripDetails(request.id)))
      .catch(err => {
        alert(`Error adding field trip note: ${JSON.stringify(err)}`)
      })
  }
}

/**
 * Delete a specific note for a field trip request.
 */
export function deleteFieldTripNote (request, noteId) {
  return function (dispatch, getState) {
    const {callTaker, otp} = getState()
    const {datastoreUrl} = otp.config
    if (sessionIsInvalid(callTaker.session)) return
    const {sessionId} = callTaker.session
    return fetch(`${datastoreUrl}/fieldtrip/deleteNote`,
      {method: 'POST', body: serialize({ noteId, sessionId })}
    )
      .then(() => dispatch(fetchFieldTripDetails(request.id)))
      .catch(err => {
        alert(`Error deleting field trip note: ${JSON.stringify(err)}`)
      })
  }
}

/**
 * Edit teacher (AKA submitter) notes for a field trip request.
 */
export function editSubmitterNotes (request, submitterNotes) {
  return function (dispatch, getState) {
    const {callTaker, otp} = getState()
    const {datastoreUrl} = otp.config
    if (sessionIsInvalid(callTaker.session)) return
    const {sessionId} = callTaker.session
    const noteData = serialize({
      notes: submitterNotes,
      requestId: request.id,
      sessionId
    })
    return fetch(`${datastoreUrl}/fieldtrip/editSubmitterNotes`,
      {method: 'POST', body: noteData}
    )
      .then(() => dispatch(fetchFieldTripDetails(request.id)))
      .catch(err => {
        alert(`Error editing submitter notes: ${JSON.stringify(err)}`)
      })
  }
}

export function saveRequestTrip (request, outbound, groupPlan) {
  return function (dispatch, getState) {
    // If plan is not valid, return before persisting trip.
    const check = checkPlanValidity(request, groupPlan)
    if (!check.isValid) return alert(check.message)
    const requestOrder = outbound ? 0 : 1
    const type = outbound ? 'outbound' : 'inbound'
    const preExistingTrip = getTripFromRequest(request, outbound)
    if (preExistingTrip) {
      const msg = `This action will overwrite a previously planned ${type} itinerary for this request. Do you wish to continue?`
      if (!confirm(msg)) return
    }
    alert(`TODO: Save trip in request order ${requestOrder}!`)
    // TODO: Enable saveTrip
    // dispatch(saveTrip(request, requestOrder))
  }
}

/**
 * @typedef {Object} ValidationCheck
 * @property {boolean} isValid - Whether the check is valid
 * @property {string} message - The message explaining why the check returned
 *   invalid.
 */

/**
 * Checks that a group plan is valid for a given request, i.e., that it occurs
 * on the requested travel date.
 * @param  request   field trip request
 * @param  groupPlan the group plan to check
 * @return {ValidationCheck}
 */
function checkPlanValidity (request, groupPlan) {
  if (groupPlan == null) {
    return {
      isValid: false,
      message: 'No active plan to save'
    }
  }

  // FIXME: add back in offset?
  const planDeparture = moment(groupPlan.earliestStartTime) // .add('hours', otp.config.timeOffset)
  const requestDate = moment(request.travelDate)

  if (
    planDeparture.date() !== requestDate.date() ||
    planDeparture.month() !== requestDate.month() ||
    planDeparture.year() !== requestDate.year()
  ) {
    return {
      isValid: false,
      message: `Planned trip date (${planDeparture.format('MM/DD/YYYY')}) is not the requested day of travel (${requestDate.format('MM/DD/YYYY')})`
    }
  }

  // FIXME More checks? E.g., origin/destination

  return { isValid: true, message: null }
}

export function planTrip (request, outbound) {
  return async function (dispatch, getState) {
    dispatch(setGroupSize(getGroupSize(request)))
    const trip = getTripFromRequest(request, outbound)
    if (!trip) {
      // Construct params from request details
      if (outbound) dispatch(planOutbound(request))
      else dispatch(planInbound(request))
    } else {
      // Populate params from saved query params
      const params = await planParamsToQueryAsync(JSON.parse(trip.queryParams))
      dispatch(setQueryParam(params, trip.id))
    }
  }
}

function planOutbound (request) {
  return async function (dispatch, getState) {
    const {config} = getState().otp
    // clearTrip()
    const locations = await planParamsToQueryAsync({
      fromPlace: request.startLocation,
      toPlace: request.endLocation
    }, config)
    const queryParams = {
      date: moment(request.travelDate).format(OTP_API_DATE_FORMAT),
      departArrive: 'ARRIVE',
      groupSize: getGroupSize(request),
      time: moment(request.arriveDestinationTime).format(OTP_API_TIME_FORMAT),
      ...locations
    }
    dispatch(setQueryParam(queryParams))
    dispatch(routingQuery())
  }
}

function planInbound (request) {
  return async function (dispatch, getState) {
    const {config} = getState().otp
    const locations = await planParamsToQueryAsync({
      fromPlace: request.endLocation,
      toPlace: request.startLocation
    }, config)
    // clearTrip()
    const queryParams = {
      date: moment(request.travelDate).format(OTP_API_DATE_FORMAT),
      departArrive: 'DEPART',
      groupSize: getGroupSize(request),
      time: moment(request.leaveDestinationTime).format(OTP_API_TIME_FORMAT),
      ...locations
    }
    dispatch(setQueryParam(queryParams))
    dispatch(routingQuery())
  }
}

/**
 * Set group size for a field trip request. Group size consists of numStudents,
 * numFreeStudents, and numChaperones.
 */
export function setRequestGroupSize (request, groupSize) {
  return function (dispatch, getState) {
    const {callTaker, otp} = getState()
    const {datastoreUrl} = otp.config
    if (sessionIsInvalid(callTaker.session)) return
    const {sessionId} = callTaker.session
    const groupSizeData = serialize({
      ...groupSize,
      requestId: request.id,
      sessionId
    })
    return fetch(`${datastoreUrl}/fieldtrip/setRequestGroupSize`,
      {method: 'POST', body: groupSizeData}
    )
      .then(() => dispatch(fetchFieldTripDetails(request.id)))
      .catch(err => {
        alert(`Error setting group size: ${JSON.stringify(err)}`)
      })
  }
}

/**
 * Set payment info for a field trip request.
 */
export function setRequestPaymentInfo (request, paymentInfo) {
  return function (dispatch, getState) {
    const {callTaker, otp} = getState()
    const {datastoreUrl} = otp.config
    if (sessionIsInvalid(callTaker.session)) return
    const {sessionId} = callTaker.session
    const paymentData = serialize({
      ...paymentInfo,
      requestId: request.id,
      sessionId
    })
    return fetch(`${datastoreUrl}/fieldtrip/setRequestPaymentInfo`,
      {method: 'POST', body: paymentData}
    )
      .then(() => dispatch(fetchFieldTripDetails(request.id)))
      .catch(err => {
        alert(`Error setting payment info: ${JSON.stringify(err)}`)
      })
  }
}

/**
 * Set field trip request status (e.g., cancelled).
 */
export function setRequestStatus (request, status) {
  return function (dispatch, getState) {
    const {callTaker, otp} = getState()
    const {datastoreUrl} = otp.config
    if (sessionIsInvalid(callTaker.session)) return
    const {sessionId} = callTaker.session
    const statusData = serialize({
      requestId: request.id,
      sessionId,
      status
    })
    return fetch(`${datastoreUrl}/fieldtrip/setRequestStatus`,
      {method: 'POST', body: statusData}
    )
      .then(() => dispatch(fetchFieldTripDetails(request.id)))
      .catch(err => {
        alert(`Error setting request status: ${JSON.stringify(err)}`)
      })
  }
}
