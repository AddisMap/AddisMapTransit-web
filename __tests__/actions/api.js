/* globals describe, expect, it, jest */

import nock from 'nock'

import {planTrip} from '../../lib/actions/api'

function timeoutPromise (ms) {
  return new Promise((resolve, reject) => {
    setTimeout(resolve, ms)
  })
}

const planTripAction = planTrip()

describe('actions > api', () => {
  it('should make a query to OTP', async () => {
    nock('http://mock-host.com')
      .get(/api\/plan/)
      .reply(200, {
        fake: 'response'
      })
      .on('request', (req, interceptor) => {
        expect(req.path).toMatchSnapshot()
      })

    const mockDispatch = jest.fn()
    planTripAction(mockDispatch, () => {
      return {
        otp: {
          config: {
            api: {
              host: 'http://mock-host.com',
              path: '/api',
              port: 80
            }
          },
          currentQuery: {
            from: {
              lat: 12,
              lon: 34
            },
            to: {
              lat: 34,
              lon: 12
            }
          },
          searches: []
        }
      }
    })

    // wait for request to complete
    await timeoutPromise(100)

    expect(mockDispatch.mock.calls).toMatchSnapshot()
  })
})
