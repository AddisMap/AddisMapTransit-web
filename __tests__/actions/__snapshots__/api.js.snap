// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions > api > planTrip should gracefully handle bad response 1`] = `
Array [
  Array [
    Object {
      "type": "PLAN_REQUEST",
    },
  ],
  Array [
    Object {
      "error": true,
      "payload": [Error: Received error from server],
      "type": "PLAN_ERROR",
    },
  ],
]
`;

exports[`actions > api > planTrip should make a query to OTP with customOtpQueryBuilder in state config 1`] = `"/api/plan?from=here&to=there"`;

exports[`actions > api > planTrip should make a query to OTP with customOtpQueryBuilder in state config 2`] = `
Array [
  Array [
    Object {
      "type": "PLAN_REQUEST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "fake": "response",
      },
      "type": "PLAN_RESPONSE",
    },
  ],
]
`;

exports[`actions > api > planTrip should make a query to OTP with customOtpQueryBuilder sent to action 1`] = `"/api/plan?from=here&to=there"`;

exports[`actions > api > planTrip should make a query to OTP with customOtpQueryBuilder sent to action 2`] = `
Array [
  Array [
    Object {
      "type": "PLAN_REQUEST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "fake": "response",
      },
      "type": "PLAN_RESPONSE",
    },
  ],
]
`;

exports[`actions > api > planTrip should make a query to OTP with default settings 1`] = `"/api/plan?arriveBy=false&fromPlace=12%2C34&showIntermediateStops=true&toPlace=34%2C12"`;

exports[`actions > api > planTrip should make a query to OTP with default settings 2`] = `
Array [
  Array [
    Object {
      "type": "PLAN_REQUEST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "fake": "response",
      },
      "type": "PLAN_RESPONSE",
    },
  ],
]
`;
